### PR TITLE
Make settings validation strict

### PR DESCRIPTION
**Pull request by s1monw**
This commit enableds strict settings validation on node startup. All settings
passed to elasticsearch either through system properties, yaml files or any other
way to pass settings must be registered and valid. Settings that are unknown ie. due to
typos or due to deprecation or removal will cause the node to NOT start up. Plugins
have to declare all their settings on the `SettingsModule#registerSetting` and settings for
plugins that are not installed must be removed.

This commit also removes the ability to specify the nodes name via `-Des.name` or just `name` in the
configuration files. The node name must be prefixed with the node prexif like `node.name: Boom`. Left over
usage of `name` will also cause startup to fail.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/settings/SettingsModule.java [hunk 1]
```diff
@@ -37,6 +38,8 @@
     private final SettingsFilter settingsFilter;
     private final Map<String, Setting<?>> clusterSettings = new HashMap<>();
     private final Map<String, Setting<?>> indexSettings = new HashMap<>();
+    private static final Predicate<String> NO_TRIBE_PREDICATE =  (s) -> s.startsWith("tribe.") == false || TribeService.TRIBE_SETTING_KEYS.contains(s);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/settings/SettingsModule.java]: This got me confused. After some research I got what we're doing (exclude settings of tribe client node). Can we maybe flips this to a positive predicate and call it `TRIBE_CLIENT_NODE_SETTING_PREDICATE` ? I think it's worth the verbosity. 

PS. good catch on the top level Tribe node setting. It's great they are validated.

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 2]
```diff
@@ -133,6 +133,11 @@
     public static final Setting<Boolean> NODE_LOCAL_SETTING = Setting.boolSetting("node.local", false, false, Setting.Scope.CLUSTER);
     public static final Setting<String> NODE_MODE_SETTING = new Setting<>("node.mode", "network", Function.identity(), false, Setting.Scope.CLUSTER);
     public static final Setting<Boolean> NODE_INGEST_SETTING = Setting.boolSetting("node.ingest", true, false, Setting.Scope.CLUSTER);
+    public static final Setting<String> NODE_NAME_SETTING = Setting.simpleString("node.name", false, Setting.Scope.CLUSTER);
+    // this sucks that folks can mistype client etc and get away with it.
+    // TODO: we should move this to node.attribute.${name} = ${value} instead.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/node/Node.java]: ++

File: core/src/main/java/org/elasticsearch/rest/action/main/RestMainAction.java [hunk 3]
```diff
@@ -76,9 +76,8 @@ public void handleRequest(final RestRequest request, RestChannel channel, final
         }
 
         builder.startObject();
-        if (settings.get("name") != null) {
-            builder.field("name", settings.get("name"));
-        }
+        assert settings.get("node.name") != null;
+        builder.field("name", settings.get("node.name"));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/rest/action/main/RestMainAction.java]: why not use the "official' setting object?

File: core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 4]
```diff
@@ -206,13 +206,13 @@ public Settings build() {
     private final ThreadContext threadContext;
 
     public ThreadPool(String name) {
-        this(Settings.builder().put("name", name).build());
+        this(Settings.builder().put("node.name", name).build());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: why not use the setting key?

File: core/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TasksIT.java [hunk 5]
```diff
@@ -66,7 +66,7 @@
     protected Settings nodeSettings(int nodeOrdinal) {
         return Settings.builder()
             .put(super.nodeSettings(nodeOrdinal))
-            .put(MockTaskManager.USE_MOCK_TASK_MANAGER, true)
+            .put(MockTaskManager.USE_MOCK_TASK_MANAGER_SETTING.getKey(), true)
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TasksIT.java]: hehe

File: core/src/test/java/org/elasticsearch/common/settings/SettingsModuleTests.java [hunk 6]
```diff
@@ -74,11 +74,40 @@ public void testTribeSetting() {
                 assertInstanceBinding(module, Settings.class, (s) -> s == settings);
                 fail();
             } catch (IllegalArgumentException ex) {
-                assertEquals("Failed to parse value [[2.0]] for setting [cluster.routing.allocation.balance.shard]", ex.getMessage());
+                assertEquals("tribe.t1 validation failed: Failed to parse value [[2.0]] for setting [cluster.routing.allocation.balance.shard]", ex.getMessage());
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/common/settings/SettingsModuleTests.java]: awesome we get the tribe prefix!

File: core/src/test/java/org/elasticsearch/gateway/RecoveryBackwardsCompatibilityIT.java [hunk 7]
```diff
@@ -43,7 +43,6 @@
     protected Settings nodeSettings(int nodeOrdinal) {
         return Settings.builder()
                 .put(super.nodeSettings(nodeOrdinal))
-                .put("action.admin.cluster.node.shutdown.delay", "10ms")
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/RecoveryBackwardsCompatibilityIT.java]: tada! this says it all..

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/DateHistogramOffsetIT.java [hunk 8]
```diff
@@ -58,6 +61,11 @@ private DateTime date(String date) {
     }
 
     @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Collections.singleton(AssertingLocalTransport.TestPlugin.class);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/DateHistogramOffsetIT.java]: curious - what triggered this?

**Comment by bleskes** [comment]
AWESOME. LGTM.  left some nitty comments. no need for another review...

**Comment by ywelsch** [comment]
I'm all ++ on this change, just a few questions:
- How will this affect users upgrading from 2.x to 3.x? How can they remove settings that were useful in 2.x but have been removed in 3.x?
- Can you add documentation for the breaking change?

**Comment by s1monw** [comment]
> Can you add documentation for the breaking change?

I will open followups for all documentation around how to register them etc.

> How will this affect users upgrading from 2.x to 3.x? How can they remove settings that were useful in 2.x but have been removed in 3.x?

settings are only validated on startup and never from the clusterstate. if you ahve something that was useful in 2.x you can just remove it from the yaml file

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/DateHistogramOffsetIT.java (hunk 8 above)
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/DateHistogramOffsetIT.java]: if you set the setting you have to register it so you can only set the setting if you ahve the plugin installed

**Comment by s1monw** [comment]
@bleskes @ywelsch I pushed updates

File: docs/reference/migration/migrate_3_0.asciidoc [hunk 9]
```diff
@@ -195,6 +195,22 @@ the parent id under the `_parent` key.
 [[breaking_30_settings_changes]]
 === Settings changes
 
+From Elasticsearch 3.0 on all settings are validated before they are applied. Node level and default index
+level settings are validated on node startup, dynamic cluster and index setting are validated before they are updated/added
+to the cluster state. Every setting must be a _known_ setting or in other words all settings must be registered with the
+node or transport client they are used with. This implies that plugins that define custom settings must register all of their
+settings during pluging loading on the `SettingsModule#registerSettings(Setting)` method.
```
Comment (clintongormley) [inline docs/reference/migration/migrate_3_0.asciidoc]: on the -> using the ?

File: docs/reference/migration/migrate_3_0.asciidoc [hunk 10]
```diff
@@ -195,6 +195,22 @@ the parent id under the `_parent` key.
...
+
+==== Node settings
+
+The `name` settings has been remove and is replaced with `node.name`. Usage of `-Dname=some_node_name` is not supported
```
Comment (clintongormley) [inline docs/reference/migration/migrate_3_0.asciidoc]: name settings -> name setting, remove -> removed, with -> by

File: docs/reference/migration/migrate_3_0.asciidoc [hunk 11]
```diff
@@ -195,6 +195,22 @@ the parent id under the `_parent` key.
...
+anymore. Scripts must be updated accordingly.
```
Comment (clintongormley) [inline docs/reference/migration/migrate_3_0.asciidoc]: Delete "Scripts must be updated accordingly."

File: docs/reference/migration/migrate_3_0.asciidoc [hunk 12]
```diff
@@ -195,6 +195,22 @@ the parent id under the `_parent` key.
...
+
+==== Transport Settings
+
+All settings with a `netty` infix have been replaced by their already existing `transport` synonyms. For instance `transport.netty.bind_host` is
+now not supported anymore and should be replaced by the superseeding setting `transport.bind_host`.
```
Comment (clintongormley) [inline docs/reference/migration/migrate_3_0.asciidoc]: now not supported anymore -> no longer supported
Delete "the superseeding setting" (or change superseeding -> superseding)

**Comment by clintongormley** [comment]
thanks for the doc changes @s1monw, i left a couple of comments

**Comment by bleskes** [comment]
thx @s1monw . LGTM.

**Pull request merged**